src/ngp_algorithms/TurbViscKsgsAlg.C

```diff
@@ -0,0 +1,66 @@
+/*------------------------------------------------------------------------*/
+/*  Copyright 2019 National Renewable Energy Laboratory.                  */
+/*  This software is released under the license detailed                  */
+/*  in the file, LICENSE, which is located in the top-level Nalu          */
+/*  directory structure                                                   */
+/*------------------------------------------------------------------------*/
+
+#include "ngp_algorithms/TurbViscKsgsAlg.h"
+#include "ngp_utils/NgpLoopUtils.h"
+#include "ngp_utils/NgpTypes.h"
+#include "Realm.h"
+#include "utils/StkHelpers.h"
+
+#include "stk_mesh/base/MetaData.hpp"
+
+namespace sierra{
+namespace nalu{
+
+TurbViscKsgsAlg::TurbViscKsgsAlg(
+  Realm &realm,
+  stk::mesh::Part *part,
+  ScalarFieldType* tke,
+  ScalarFieldType* density,
+  ScalarFieldType* tvisc,
+  ScalarFieldType* dualNodalVolume
+) : Algorithm(realm, part),
+    tviscField_(tvisc),
+    tke_(tke->mesh_meta_data_ordinal()),
+    density_(density->mesh_meta_data_ordinal()),
+    tvisc_(tvisc->mesh_meta_data_ordinal()),
+    dualNodalVolume_(dualNodalVolume->mesh_meta_data_ordinal()),
+    cmuEps_(realm.get_turb_model_constant(TM_cmuEps))
+{}
+
+void
+TurbViscKsgsAlg::execute()
+{
+  using Traits = nalu_ngp::NGPMeshTraits<ngp::Mesh>;
+
+  const auto& meta = realm_.meta_data();
+
+  stk::mesh::Selector sel = (
+    meta.locally_owned_part() | meta.globally_shared_part())
+    & stk::mesh::selectField(*tviscField_);
+
+  const auto& meshInfo = realm_.mesh_info();
+  const auto ngpMesh = meshInfo.ngp_mesh();
+  const auto& fieldMgr = meshInfo.ngp_field_manager();
+  const auto tke = fieldMgr.get_field<double>(tke_);
+  const auto density = fieldMgr.get_field<double>(density_);
+  const auto dualNodalVolume = fieldMgr.get_field<double>(dualNodalVolume_);
+  auto tvisc = fieldMgr.get_field<double>(tvisc_);
+
+  const DblType cmuEps = cmuEps_;
+  const DblType invNdim = 1.0 / meta.spatial_dimension();
+
+  nalu_ngp::run_entity_algorithm(
+    ngpMesh, stk::topology::NODE_RANK, sel,
+    KOKKOS_LAMBDA(const Traits::MeshIndex& meshIdx) {
+      const DblType filter = std::pow(dualNodalVolume.get(meshIdx, 0), invNdim);
```

Comment:
Is there a benefit to using `std::sqrt` and `std::cbrt` here instead of generic `std::pow`? We did that in a Kernel 

https://github.com/Exawind/nalu-wind/blob/3b9098b0965d421cf97d95f7b1cc8fe70412f41c/src/kernel/TurbKineticEnergyKsgsSrcElemKernel.C#L89-L91

Reply:
Does this run on GPU? I didn't think std:: functions would work on GPU, instead we need to use the stk::math:: versions.

Reply:
@alanw0 `stk::math::pow` in turn calls `std::pow` anyway, so I don't think that will be an issue on device. The pitfall is that if we ever switch `DblType` to `stk::simd::Double` then the code would break on `simd`.

https://github.com/trilinos/Trilinos/blob/ae1e3b063186d08c9dade332701e92257a6fe24f/packages/stk/stk_math/stk_math/StkMath.hpp#L79-L81